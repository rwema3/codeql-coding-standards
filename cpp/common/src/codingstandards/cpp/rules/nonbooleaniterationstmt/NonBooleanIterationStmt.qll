/**
 * Provides a library which includes a `problems` predicate for reporting....
 */

import cpp
import codingstandards.cpp.Customizations
import codingstandards.cpp.Exclusions

abstract class NonBooleanIterationStmtSharedQuery extends Query { }

Query getQuery() { result instanceof NonBooleanIterationStmtSharedQuery }

query predicate problems(Loop loopStmt, string message) {
  not isExcluded(loopStmt, getQuery()) and
  exists(Expr condition, Type explicitConversionType |
    condition = loopStmt.getCondition() and
    explicitConversionType = condition.getExplicitlyConverted().getType().getUnspecifiedType() and
    not explicitConversionType instanceof BoolType and
    message = "Iteration condition has non boolean type " + explicitConversionType + "."
  )
}
