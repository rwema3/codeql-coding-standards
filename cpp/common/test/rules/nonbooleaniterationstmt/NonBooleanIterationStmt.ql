// GENERATED FILE - DO NOT MODIFY
import codingstandards.cpp.rules.nonbooleaniterationstmt.NonBooleanIterationStmt

class TestFileQuery extends NonBooleanIterationStmtSharedQuery, TestQuery { }
