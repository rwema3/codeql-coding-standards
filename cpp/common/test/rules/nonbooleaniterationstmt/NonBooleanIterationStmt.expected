| test.cpp:7:3:9:3 | for(...;...;...) ... | Iteration condition has non boolean type int. |
| test.cpp:11:3:13:3 | while (...) ... | Iteration condition has non boolean type int. |
